cargo/cargo_build_script.bzl

```diff
@@ -0,0 +1,108 @@
+load("@io_bazel_rules_rust//rust:private/rustc.bzl", "BuildInfo", "rustc_compile_action")
+load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
+
+def _build_script_impl(ctx):
+    toolchain = find_toolchain(ctx)
+    out_dir = ctx.actions.declare_directory(ctx.label.name + ".out_dir")
+    env_out = ctx.actions.declare_file(ctx.label.name + ".env")
+    flags_out = ctx.actions.declare_file(ctx.label.name + ".flags")
+    env = {
+        "CARGO_MANIFEST_DIR": ctx.build_file_path.rsplit("/", 1)[0],
+        "RUSTC": toolchain.rustc.path,
+        "TARGET": toolchain.target_triple,
+        "OUT_DIR": out_dir.path,
+        "BINARY_PATH": ctx.executable.script.path,
+    }
+
+    for f in ctx.attr.crate_features:
+        env["CARGO_FEATURE_" + f.upper().replace("-", "_")] = "1"
+
+    ctx.actions.run(
+        executable = ctx.executable._cargo_build_script_runner,
+        arguments = [ctx.executable.script.path, env_out.path, flags_out.path],
+        outputs = [out_dir, env_out, flags_out],
+        tools = [ctx.executable.script, ctx.executable._cargo_build_script_runner],
+        inputs = ctx.files.srcs + [ctx.executable.script, toolchain.rustc],
+        mnemonic = "CargoBuildScriptRun",
+        env = env,
+    )
+
+    return [
+        BuildInfo(
+            out_dir = out_dir,
+            rustc_env = env_out,
+            flags = flags_out,
+        ),
+    ]
+
+cargo_build_script_run = rule(
+    _build_script_impl,
+    attrs = {
+        "script": attr.label(
+            executable = True,
+            allow_files = True,
+            mandatory = True,
+            cfg = "host",
+            doc = "The binary script to run, generally a rust_binary target. ",
+        ),
+        "crate_features": attr.string_list(doc = "The list of rust features that the build script should consider activated."),
+        "srcs": attr.label_list(allow_files = True, doc = "A list of files the scripts needs to access, generally the sources of the crate for which the build script is working"),
+        "_cargo_build_script_runner": attr.label(
+            executable = True,
+            allow_files = True,
+            default = Label("//cargo/cargo_build_script_runner:cargo_build_script_runner"),
+            cfg = "host",
+        ),
+    },
+    toolchains = [
+        "@io_bazel_rules_rust//rust:toolchain",
+    ],
+    doc = """
+Execute a rust build script to generate build attributes.
+
+Example:
+
+Suppose you have a crate with a cargo build script `build.rs`:
+
+```
+[workspace]/
+    hello_lib/
+        BUILD
+        build.rs
+        src/
+            lib.rs
+```
+
+Then you want to use the build script in the following:
+
+`hello_lib/BUILD`:
+```python
+package(default_visibility = ["//visibility:public"])
+
+load("@io_bazel_rules_rust//rust:rust.bzl", "rust_binary", "rust_library")
+load("@io_bazel_rules_rust//cargo:cargo_build_script.bzl", "cargo_build_script")
+
+rust_binary(
```

Comment:
What are your thoughts on merging the "cargo_build_script_run" and "rust_binary" into one target?

```
cargo_build_script_run(
  name = "build_script",
  srcs = ["build.rs"]
)

rust_library(
  ...
  deps = [":build_script"]
)
```

Couple of reasons why I'm wondering about this:
- Does the current usage of "srcs" in `cargo_build_script_run` -- where the library files are passed -- make sense? I haven't seen build script code that actually depends on corresponding library files. If we wanted to expose files to the build script, perhaps `data` dependencies would be more explicit, as these could be any files, not necessarily source?
- Is there a case where it makes sense to specify `cargo_build_script_run` without the build script binary, or vice versa? I'm struggling to see an example of this case.

Reply:
* In google/starlark-rust I need to read the sources (in that case the grammar). There are a bunch of preprocessor working on the code themselve too (although those one are mostly macro now).
* I don't know about a case, the logic is separated here because it would require to recreate the logic of the `rust_binary`. I would rather create a shortcut macro rather than making it a separate rules.

Reply:
I don't understand the first response, about "in google/starlark-rust".  Can you clarify? Actually I don't follow the second one either.

I'm inclined to favor a single target that builds and runs the build script, rather than separate build and run targets. If that's possible.

Reply:
Rephrasing the first one: In starlark-rust, the build script needs to reads some sources, hence the need for a sources (maybe srcs is not the best place for it though, data would be more logical). There is more case of generative code that want to read the sources input.

For the second point, I am just saying that I don't have a use case. This approach is made for simplicity. I am up for exposing only one rule but I would advise to make it a macro (basically keeping the underlying logic identical).

I'll go ahead and do the change so you can see how you feel about it.

Reply:
Change uploaded.